### DISCUSSION TITLE
对 `plugins/data` 的源码逻辑感到困惑，特来询问

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#3155] [综合讨论 / General]
我最近研究了 `plugins/data` 的源码，发现其中有很多令人困惑、感到冲突的地方，我猜测应该是什么地方写错了。本来打算直接改了，但写了一半感觉没理解原作者的意思，所以来问问。

1. 一个 provider 是否执行且只执行一次？
   如果是，那执行一次后是否可以直接丢弃该 provider
   如果不是，应该哪些时机执行 provider
3. 在 register 前的 addData 是否丢弃
4. getData 是否只真正执行一次，其后的调用都使用缓存起来的数据？如果是，那 getData 后的 addData 是否直接丢弃
5. 是否假设 register 一定在 getData 前执行过一次，即由调用方来保证调用顺序的正确性
6. 是否仅在 getData 时才真正执行保存起来的 provider
7. 对三个函数的调用时机是否有限制？比如 addData 只能在插件的 setup 中调用，以保证所有 addData 在 register 或 getData 前被调用？

**Comment by the1812** [comment]
1. 是的, 数据是单向的, 丢弃了应该也没问题, 虽然我现在没丢弃.
2. 不丢弃, Plugins API 要做到顺序无关, `add add register get` 和 `register add add get` 应该产生一致的数据.
3. 可以多次, 后续如果还有 addData, getData 调用获取到的仍然是最新的数据. 最新的数据如何获取看调用方, 可以再次调用 getData, 也可以把一开始拿到的 data 直接丢进 Vue data 里做响应式.
4. 通常是这样, 也提供了连续调用的 registerAndGetData. 但不强制要求, 如果没 register 过 getData 只是拿到空数据而已.
5. 是的, 这里是惰性求值的, 因为数据不一定真的会被 getData, 比如你提供了快捷键支持, 但用户根本没装快捷键扩展.
6. 没有限制, 这套 API 是通用的, 在任何地方都可以使用, 在组件里或是普通的函数里都可以. 顺序的问题依然还是第 2 条解释过了.

**Comment by timongh** [comment]
既然这三个函数的执行顺序得不到保证，那么组件/插件的效果怎么得到保证呢？比如某个组件/插件通过 key 接受其他部分的扩展。但是接受方却在提供方之前启动。那提供方的扩展岂不是直接失效。

**Reply by the1812** [reply]
就是上面第 3 条的场景, "是否失效" 也就是后来的 addData 有没有被用起来, 这取决于接收方怎么去 getData.
